### ISSUE TITLE
java.util.Collections$EmptyMap cannot be cast to org.elasticsearch.hadoop.mr.LinkedMapWritable

### ISSUE THREAD

**Issue by boliza** [elastic/elasticsearch-hadoop#560]
I use es 1.2.1 , hadoop2.6 , my schema is 

``` json
{
  "vid": {
    "mappings": {
      "vid_vertex": {
        "_ttl": {
          "enabled": true
        },
        "properties": {
          "createdOn": {
            "type": "long"
          },
          "domain": {
            "type": "string",
            "index": "not_analyzed"
          },
          "oid": {
            "type": "string",
            "index": "not_analyzed"
          },
          "password": {
            "type": "string",
            "index": "not_analyzed"
          },
          "pdomain": {
            "type": "string",
            "index": "not_analyzed"
          },
          "vid": {
            "type": "string",
            "index": "not_analyzed"
          },
          "vidAddress": {
            "type": "string",
            "analyzer": "jieba"
          },
          "vidCategory": {
            "type": "string",
            "index": "not_analyzed"
          },
          "vidDescription": {
            "type": "string",
            "analyzer": "jieba"
          },
          "vidName": {
            "type": "string",
            "analyzer": "jieba"
          },
          "vidNickName": {
            "type": "string",
            "analyzer": "jieba"
          },
          "vidType": {
            "type": "string",
            "index": "not_analyzed"
          }
        }
      }
    }
  }
}
```

There is MR Traversal my code

``` java
        getConf().set("es.nodes", "es1:9200");
        getConf().set("es.resource", "vid/vid_vertex");
        getConf().set("es.query", "{\n" +
            "  \"fields\": [\n" +
            "    \"vid\"\n" +
            "  ],\n" +
            "  \"query\": {\n" +
            "    \"term\": {\n" +
            "      \"vidType\": \"102\"\n" +
            "    }\n" +
            "  }\n" +
            "}");

        Job job = Job.getInstance(getConf());
        job.setJobName("Traversal Vid from ES");
        job.setJarByClass(EsMr.class);

        job.setInputFormatClass(EsInputFormat.class);
        job.setMapperClass(EsMapper.class);
        job.setMapOutputKeyClass(Text.class);
        job.setMapOutputValueClass(LinkedMapWritable.class);

        job.setReducerClass(EsReducer.class);
        job.setOutputKeyClass(Text.class);
        job.setOutputValueClass(Text.class);
        job.setOutputFormatClass(TextOutputFormat.class);
```

Mapper and Reducer is

``` java
public static class EsMapper extends Mapper<Text, LinkedMapWritable, Text, LinkedMapWritable> {

        @Override
        protected void map(Text key, LinkedMapWritable value, Context context) throws IOException, InterruptedException {
            context.write(key, value);
        }

    }

    public static class EsReducer extends Reducer<Text, LinkedMapWritable, Text, Text> {

        private Text vid = new Text("vid");
        private Text outputValue = new Text();

        @Override
        public void reduce(Text key, Iterable<LinkedMapWritable> values, Context context) throws IOException, InterruptedException {
            for (LinkedMapWritable value : values) {
                ArrayWritable idArray = (ArrayWritable) value.get(vid);
                if (idArray == null || idArray.get()[0] == null) {
                    return;
                }
                outputValue.set(idArray.get()[0].toString().substring(4) + "\t" + "website");
                context.write((Text) idArray.get()[0], outputValue);
            }
        }
    }
```

After run this mr ,hadoop report an ClassCastException.

Could someone help to resolve this???

**Comment by costin** [comment]
What version of es-hadoop are you using and what's the stracktrace?

**Comment by boliza** [comment]
dependency is 

```xml       
       <dependency>
            <groupId>org.elasticsearch</groupId>
            <artifactId>elasticsearch-hadoop-mr</artifactId>
            <version>2.0.2</version>
        </dependency>
```

exception stracktrace is 

```java
Error: java.lang.ClassCastException: java.util.Collections$EmptyMap cannot be cast to org.elasticsearch.hadoop.mr.LinkedMapWritable
	at com.antfact.nest.graph.traversal.EsMr$EsMapper.map(EsMr.java:74)
	at org.apache.hadoop.mapreduce.Mapper.run(Mapper.java:145)
	at org.apache.hadoop.mapred.MapTask.runNewMapper(MapTask.java:784)
	at org.apache.hadoop.mapred.MapTask.run(MapTask.java:341)
	at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:163)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:415)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1628)
	at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:158)
```

**Comment by costin** [comment]
Can you post your `EsMR` class?
The issue seems to be caused by your exception of `LinkedMapWritable`.
That's a class that is used by the connector and it's typically used by
user code when reading data from Hadoop. Better yet, instead of referring
to it, one can simply use the `Map` interface instead.

In your case however it looks like you are using this type in your mapper
which means you are expecting that the types that will be written to
Elasticsearch (through the connector) will fit it. Which typically is not
the case.
In other words, try to modify your `Mapper` to `Mapper<Text, Map, Text,
Map>` - while generics sound like a good idea, in case of ETL due to the
fact that data tends not fit into one schema, it's better to have a loose
signature instead of a precise one.

**Comment by boliza** [comment]
sorry to reply to late,my EsMr is 

```java
public class EsMr extends Configured implements Tool {

    private static final String OUTPUT_PATH_FORMAT = "/eefung/graph/es/traversal/vid/%s";

    public static void main(String[] args) throws Exception {
        Configuration conf = new Configuration();
        ToolRunner.run(conf, new EsMr(), args);
    }

    @Override
    public int run(String[] args) throws Exception {

        if (args == null || args.length < 2) {
            System.out.println("must 2 args,like \r" +
                                   "10 (number of reduce)\r" +
                                   "20150929 (output path)\r");
            return 0;
        }

        getConf().set("es.nodes", "es1:9200");
        getConf().set("es.resource", "vid/vid_vertex");
        getConf().set("es.query", "{\n" +
            "  \"fields\": [\n" +
            "    \"vid\"\n" +
            "  ],\n" +
            "  \"query\": {\n" +
            "    \"term\": {\n" +
            "      \"vidType\": \"102\"\n" +
            "    }\n" +
            "  }\n" +
            "}");

        Job job = Job.getInstance(getConf());
        job.setJobName("Traversal Vid from ES");
        job.setJarByClass(EsMr.class);

        job.setInputFormatClass(EsInputFormat.class);
        job.setMapperClass(EsMapper.class);
        job.setMapOutputKeyClass(Text.class);
        job.setMapOutputValueClass(LinkedMapWritable.class);

        job.setReducerClass(EsReducer.class);
        job.setOutputKeyClass(Text.class);
        job.setOutputValueClass(Text.class);
        job.setOutputFormatClass(TextOutputFormat.class);

        job.setNumReduceTasks(Integer.valueOf(args[0]));

        FileOutputFormat.setOutputPath(job, new Path(String.format(OUTPUT_PATH_FORMAT, args[1])));

        return job.waitForCompletion(true) ? 0 : 1;
    }

    public static class EsMapper extends Mapper<Text, LinkedMapWritable, Text, LinkedMapWritable> {

        @Override
        protected void map(Text key, LinkedMapWritable value, Context context) throws IOException, InterruptedException {
            context.write(key, value);
        }

    }

    public static class EsReducer extends Reducer<Text, LinkedMapWritable, Text, Text> {

        private Text vid = new Text("vid");
        private Text outputValue = new Text();

        @Override
        public void reduce(Text key, Iterable<LinkedMapWritable> values, Context context) throws IOException, InterruptedException {
            for (LinkedMapWritable value : values) {
                ArrayWritable idArray = (ArrayWritable) value.get(vid);
                if (idArray == null || idArray.get()[0] == null) {
                    return;
                }
                outputValue.set(idArray.get()[0].toString().substring(4) + "\t" + "website");
                context.write((Text) idArray.get()[0], outputValue);
            }
        }
    }

}
```

**Comment by boliza** [comment]
after change from LinkedMapWritable to Map ,also an exception ,statcktrace is 

```java
2015-09-29 21:46:41,742 WARN [main] org.apache.hadoop.mapred.MapTask: Unable to initialize MapOutputCollector org.apache.hadoop.mapred.MapTask$MapOutputBuffer
java.lang.NullPointerException
	at org.apache.hadoop.mapred.MapTask$MapOutputBuffer.init(MapTask.java:1008)
	at org.apache.hadoop.mapred.MapTask.createSortingCollector(MapTask.java:401)
	at org.apache.hadoop.mapred.MapTask.access$100(MapTask.java:81)
	at org.apache.hadoop.mapred.MapTask$NewOutputCollector.<init>(MapTask.java:695)
	at org.apache.hadoop.mapred.MapTask.runNewMapper(MapTask.java:767)
	at org.apache.hadoop.mapred.MapTask.run(MapTask.java:341)
	at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:163)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:415)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1628)
	at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:158)
2015-09-29 21:46:41,753 WARN [main] org.apache.hadoop.mapred.YarnChild: Exception running child : java.io.IOException: Unable to initialize any output collector
	at org.apache.hadoop.mapred.MapTask.createSortingCollector(MapTask.java:412)
	at org.apache.hadoop.mapred.MapTask.access$100(MapTask.java:81)
	at org.apache.hadoop.mapred.MapTask$NewOutputCollector.<init>(MapTask.java:695)
	at org.apache.hadoop.mapred.MapTask.runNewMapper(MapTask.java:767)
	at org.apache.hadoop.mapred.MapTask.run(MapTask.java:341)
	at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:163)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:415)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1628)
	at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:158)
```

**Comment by boliza** [comment]
Another info : I run this EsMr in hadoop2.4 perfect a few months ago

**Comment by costin** [comment]
Looks like you bumped into a bug in Hadoop - probably the Mapper/Reducer
signature is tripping of initialization.
I'll have to look a bit deeper to figure out what causes the exception.

**Comment by boliza** [comment]
Thank u very much!!!

**Comment by costin** [comment]
@boliza I tried several tests but I can't reproduce the error. Can you tell me your exact Hadoop version (are you using Vanilla or a certain distro)?

Thanks,

**Comment by boliza** [comment]
HDP 2.6 .Sorry for reply so mush later.

**Comment by costin** [comment]
I assume you mean Hadoop (not HDP the distro) 2.6. What's the patch number (that is the x in 2.6.x)?

**Comment by boliza** [comment]
@costin my mistake. I use Hortonworks Data Platform 2.3

**Comment by jbaiera** [comment]
This issue is quite old with no further conversation. I am closing this as an inactive ticket. If the issue is still present in more recent versions of the connector, please open a new issue.
